### PR TITLE
Issue 12421 - Allow simpler syntax for lambda template declarations

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=12421

Corresponding compiler change: https://github.com/D-Programming-Language/dmd/pull/3638

### CODE REVIEW

File: declaration.dd [hunk 1]
```diff
@@ -325,8 +328,12 @@ $(GNAME AliasDeclaration):
     $(D alias) $(I AliasDeclarationX) $(D ;)
 
 $(GNAME AliasDeclarationX):
+    $(GLINK AliasDeclarationY)
+    $(I AliasDeclarationX) $(D ,) $(GLINK AliasDeclarationY)
+
+$(GNAME AliasDeclarationY):
     $(I Identifier) $(GLINK2 template, TemplateParameters)$(OPT) $(D =) $(GLINK StorageClasses)$(OPT) $(GLINK Type)
-    $(I AliasDeclarationX) $(D ,) $(I Identifier) $(GLINK2 template, TemplateParameters)$(OPT) $(D =) $(GLINK StorageClasses)$(OPT) $(GLINK Type)
+    $(I Identifier) $(GLINK2 template, TemplateParameters)$(OPT) $(D =) $(GLINK2 expression, FunctionLiteral)
```
Comment (MartinNowak) [inline declaration.dd]: What are the optional template arguments for?

Comment (9rnsr) [inline declaration.dd]: To maximize the power of new syntax.

A template lambda e.g. `x => x + 1` is equivalent with a normal template function `auto __lambda(T)(T x) { return x + 1; }`, but it still cannot have explicit template parameters. By using the optional template parameters on alias declaration, we can have complete alternative syntax against normal template function declarations.

```d
alias foo(alias pred) = x => pred(x) + 1;
// is equivalent with:
tempalte foo(alias pred) { auto foo(T)(T x) { return pred(x) + 1; } }
```

**Comment by CyberShadow** [comment]
LGTM but some examples would be nice.

**Comment by CyberShadow** [comment]
@9rnsr Let me know if you want to add some examples or would rather have this merged as-is.

**Comment by 9rnsr** [comment]
I'll reopen this PR to stable branch, with addition of changelog updating.

**Comment by 9rnsr** [comment]
Done: #1147

**Pull request closed without merging**